### ISSUE TITLE
[Win32 quadmath] isfinite() does not accommodate __float128

### ISSUE THREAD

**Issue by sisyphus** [houseabsolute/DateTime.pm#139]
Hi,
On quadmath builds on Windows, <code>fpclassify()</code>, which is used by <code>isfinite()</code>, does not handle the <code>__float128</code> data type.

As a consequence, all of the DateTime-1.59 test scripts die without providing any output at all.

This patch (below) fixes the problem by casting the problematic NVs to long double - though this is done only if both <code>WIN32</code> && <code>USE_QUADMATH</code> are defined.
The maximum finite long double value is <code>1.18973149535723176502e4932</code>, whereas the maximum finite __float128 value is <code>1.18973149535723176508575932662800702e4932</code>
We can see that there will be some finite __float128 values that will cast to a long double of INF - but I doubt that will be of any consequence, in practice.

According to a comment in DateTime.xs, the "file is generated by tools/leap_seconds_header.pl".
If I need to instead provide a patch for tools/leap_seconds_header.pl, then please tell me where it's located and I'll take a look.

Cheers,
Rob

[DT_xs_diff.txt](https://github.com/houseabsolute/DateTime.pm/files/12060103/DT_xs_diff.txt)

**Comment by sisyphus** [comment]
Better would be to use Perl_isfinite when both <code>WIN32</code> && <code>USE_QUADMATH</code> are defined.
See revised patch below.

One wonders whether we could just replace the <code>isfinite()</code> calls in DateTime.xs with <code>Perl_isfinite()</code> calls - and have that apply to all perls, irrespective of OS and architecture.
That might also then enable the removal of some <code>#define</code> and <code>#ifdef</code> calls.
Then again, we might find that opens up  a can of worms that would better have remained closed ;-)

Cheers,
Rob

[DT_xs_diff2.txt](https://github.com/houseabsolute/DateTime.pm/files/12060301/DT_xs_diff2.txt)

**Comment by autarch** [comment]
Thanks for this. I think the even simpler fix is to simply use `Perl_isfinite` on all platforms, unconditionally. I made a branch to try that out.

**Comment by sisyphus** [comment]
LGTM ;-)
The origin/isfinite branch works fine for me  for both 32-bit and 64-bit builds of perl-5.38.0.
For each of those 2 architectures, I've tested on 'double', long double' and '__float128' builds.
I also tested the 64-bit perl-5.36.0 '__float128' build, which was fine, too.

FYI, the quadmath (__float128) option became available on Windows starting with perl-5.34.0.

Cheers,
Rob

**Comment by autarch** [comment]
I released a 1.60-TRIAL release with this change. I'll wait a month to get some CPAN testers results. If it looks good then I'll do a non-TRIAL release.

**Comment by sisyphus** [comment]
Cool !!
This trial release builds and tests fine for me on Windows 11 build of -Dusequadmath perl-5.38.0.
(I don't think there are any cpan-testers who test Windows -Dusequadmath builds.)

I'll wait until the non-TRIAL release before closing this .... just in case something relevant pops up in the meantime.
Or, you can close it now, if you prefer.

Cheers,
Rob

**Comment by autarch** [comment]
I'll leave it open until the non-TRIAL release. Thanks.

**Comment by autarch** [comment]
CPAN Testers hows some failures on Windows with older Perl versions (<= 5.20) - http://matrix.cpantesters.org/?dist=DateTime+1.60-TRIAL

They all look like this:

```
C:\"Strawberry\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"DateTime\", 'DLBASE' => 'DateTime', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' => {  }, 'DL_VARS' => []);"
g++ DateTime.def -o blib\arch\auto\DateTime\DateTime.dll -mdll -s -L"C:\Strawberry\perl\lib\CORE" -L"C:\Strawberry\c\lib" DateTime.o   "C:\Strawberry\perl\lib\CORE\libperl512.a" -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32 -Wl,--enable-auto-image-base
DateTime.o:DateTime.c:(.text+0x1ce5): undefined reference to `Perl_is_inf'
DateTime.o:DateTime.c:(.text+0x1d03): undefined reference to `Perl_is_nan'
DateTime.o:DateTime.c:(.text+0x1d24): undefined reference to `Perl_is_inf'
DateTime.o:DateTime.c:(.text+0x1d45): undefined reference to `Perl_is_nan'
DateTime.o:DateTime.c:(.text+0x24df): undefined reference to `Perl_is_inf'
DateTime.o:DateTime.c:(.text+0x24fd): undefined reference to `Perl_is_nan'
DateTime.o:DateTime.c:(.text+0x251b): undefined reference to `Perl_is_inf'
DateTime.o:DateTime.c:(.text+0x2539): undefined reference to `Perl_is_nan'
collect2: ld returned 1 exit status
dmake.EXE:  Error code 129, while making 'blib\arch\auto\DateTime\DateTime.dll'
```

I'm guessing that `Perl_isfinite` doesn't exist or is somehow broken with those older Perls on Windows. I think the easiest thing is probably just go back to the older code on platforms where `Perl_isfinite` doesn't work. I'll do another TRIAL release with that change soon.

**Comment by sisyphus** [comment]
On Sun, Sep 3, 2023 at 1:20 AM Dave Rolsky ***@***.***> wrote:

> CPAN Testers hows some failures on Windows with older Perl versions (<=
> 5.20) - http://matrix.cpantesters.org/?dist=DateTime+1.60-TRIAL
>
> I'm guessing that Perl_isfinite doesn't exist or is somehow broken with
> those older Perls on Windows. I think the easiest thing is probably just go
> back to the older code on platforms where Perl_isfinite doesn't work.
> I'll do another TRIAL release with that change soon.
>
Your proposal sounds fine to me.
The first stable release of perl that enabled quadmath builds on MS Windows
was 5.34.0.

Cheers,
Rob

**Comment by autarch** [comment]
I saw the same CPAN Testers failures again with 1.61. This time I went and tried the new DateTime out on my Windows machine. I installed both 32- and 64-bit versions of 5.20.3 and could not replicate this problem. I'm wondering if this is just an issue with someone's testing machine. It seems like the compiler thinks the `Perl_is_inf` function exists but the linker doesn't. Very weird.

**Comment by sisyphus** [comment]
On Windows 11, I see the reported Date-Tome-1.60 test failures with:
https://strawberryperl.com/download/5.20.3.3/strawberry-perl-5.20.3.3-64bit-portable.zip

Also, running this Inline::C script:
```
### try.pl ###
use strict;
use warnings;

use Inline C => Config =>
  BUILD_NOISY => 1,
;

use Inline C => <<'EOC';

void foo() {
  if(Perl_is_Inf(17)) printf("Found Perl_is_inf()\n");
  else printf("Still found Perl_is_inf()\n");
}

EOC

foo();
```
on that perl-5.20.3, I see the error:
<code>
try_pl_b45a.o:try_pl_b45a.c:(.text+0xa): undefined reference to `Perl_is_Inf'
collect2.exe: error: ld returned 1 exit status
</code>
and I think that's a pretty strong indication that <code>Perl_is_inf</code> is not, in fact, available.
But there's actually no complaint about that function during compilation - with either DateTime-1.60 or that Inline::C script.

I guess that implies that the compiler can see that <code>Perl_is_inf</code> has been protptyped, but the function cannot be linked in for some reason. (Maybe, at that tijme, it simply had not been marked for export - which would be an issue on Windows, but not on linux  type systems.)

Cheers,
Rob

**Comment by autarch** [comment]
Fixed in 1.63.
